### DISCUSSION TITLE
Enumeration

### DISCUSSION THREAD

**Discussion by cemoktra** [SeaQL/sea-orm#196] [Ideas]
Generating the code, is there anything planned about enumerations? Getting enums from db or allowing custom enum mapping?

**Comment by billy1624** [comment]
Do you mean mapping between Postgres enum and Rust enum?

**Reply by cemoktra** [reply]
Yes for example.

**Comment by tyt2y3** [comment]
Thank you for using the discussion section.
There are several stages this feature will be approached.

1) Rust enum support (but serialize to string or integer in SQL)
2) Mapping to a corresponding SQL enum type
3) Schema discovery & code generation

I think 1 is already possible currently, but we have not had a documented usage.

**Reply by tyt2y3** [reply]
@nappa85 in issue https://github.com/SeaQL/sea-orm/issues/107 I see that you were trying to make a custom enum.
I suppose it is already solved in `0.2`.
Can you perhaps also share your example model and usage bits?
I am thinking, whether it is a good idea to add a `ActiveEnum` value type, where it is actually a `Box<dyn EnumTrait>`.

**Reply by nappa85** [reply]
My Enum impl is simply an enum that decodes from a String, e.g.:
```rust
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum Boolean {
    Yes,
    No,
}

impl Boolean {
    pub const fn as_str(&self) -> &'static str {
        match self {
            Boolean::Yes => "Y",
            Boolean::No => "N",
        }
    }
}

impl Default for Boolean {
    fn default() -> Self {
        Boolean::No
    }
}

impl<'a> TryFrom<&'a str> for Boolean {
    type Error = String;
    fn try_from(c: &'a str) -> Result<Self, Self::Error> {
        match c {
            "Y" => Ok(Boolean::Yes),
            "N" => Ok(Boolean::No),
            _ => Err(format!("Unrecognized boolean value {}", c)),
        }
    }
}

impl From<Boolean> for Value {
    fn from(b: Boolean) -> Self {
        Value::from(b.as_str())
    }
}

impl Nullable for Boolean {
    fn null() -> Value {
        Value::String(None)
    }
}

impl ValueType for Boolean {
    fn try_from(v: Value) -> Result<Self, ValueTypeErr> {
        match v {
            Value::String(Some(s)) => s.as_str().try_into().map_err(|e| {
                error!("{}", e);
                ValueTypeErr
            }),
            _ => {
                error!("type error");
                Err(ValueTypeErr)
            },
        }
    }
    fn type_name() -> String {
        "Boolean".to_owned()
    }
}

impl TryGetable for Boolean {
    fn try_get(res: &QueryResult, pre: &str, col: &str) -> Result<Self, sea_orm::TryGetError> {
        let s = String::try_get(res, pre, col)?;
        s.as_str().try_into().map_err(|e| sea_orm::TryGetError::DbErr(DbErr::Query(e)))
    }
}
```

**Reply by tyt2y3** [reply]
Thank you so much for the example.
We should think about how to eliminate these boilerplate, and then add it to our test suite!

**Reply by Stumblinbear** [reply]
Something to note is this wouldn't work if used in a primary key; it would also need to implement TryFromU64 (which it shouldn't need). I just ran into this issue.

**Reply by nappa85** [reply]
Why use an enum as auto_increment primary key? As far as I know u64 is mandatory only when auto_increment is on

**Reply by Stumblinbear** [reply]
It's within a composite key (all primary key fields are set to auto_increment false).

**Reply by nappa85** [reply]
Can you make an example?

**Comment by tyt2y3** [comment]
We have two (or three) possibilities here,
1) Enum backed by a one char string (as shown above)
2) Enum backed by a TinyInteger (hence 1 byte)
3) Enum backed by a database enum (though their definition is quite different between MySQL and Postgres)

**Reply by nappa85** [reply]
I would use a macro and specify Value variant on the attribute, something like
```rust
#[derive(OurEnumMacro)]
enum Foo {
#[sea_orm(from = "String(\"A\")")]
A,
}
```
I admit is quite ugly with the backslashes, maybe it's possible to omit external quotes...

**Reply by zexa** [reply]
I would very much prefer if SeaOrm could map from PostgreSQL custom types Or MySQL enums instead of using a From<&str> approach. 

Sometimes as an engineer, you have to turn to hacky solutions (such as inserting or updating the database directly). In such cases having the database itself ensure that no invalid states are provided is beneficial and in line with rust's strict type philosophy where the compiler slaps you in the face when you're trying to do something that won't end up well.

If an engineer ends up inserting an enum value that's not actually valid in the software we might get panics  and then I'll have to fish out the exact entries in the database that are causing issues. That idea doesn't ring well with me.

**Reply by billy1624** [reply]
Please check #261 it support native postgres enum. While working towards mysql enum in coming days.

**Comment by cemoktra** [comment]
For general enumerations (not defined in SQL) an approach might be to offer a config file for code generation.

The config file could contain mapping for types in specific tables like:

```
[tab_example]
column_name = "module::MyEnum"
```

Of course `MyEnum` must implement some specific trait but it might be a way to have a general type mapping
